README.md
# Timetoc

Timetoc is a command-line tool for tracking work time in the timetac time tracking system. The tool allows users to add work days using the timetac API.

# Installation

To install the Timetoc Work Tracker tool, run the following command:

```console
$ pipx install git+https://github.com/guilhermeprokisch/timetoc.git
```

# Configuration

Add `TIME_TRACK_BASE_URL` as enviroment variable for the crawler works

```console
$ export TIME_TRACK_BASE_URL="https://some-timetac-login-page.com
```

# Usage



The following options are available:
```
  --day or -d: The date of the work day in the format yyyy-mm-dd.
  --day-str or -ds: The date of the work day in natural language. E.g 'today, yesterday'.
  --start or -s: The start time of the work day in the format hh:mm.
  --finish or -f: The finish time of the work day in the format hh:mm.
  --break-start or -bs: The start time of the break in the work day in the format hh:mm.
  --break-finish or -bf: The finish time of the break in the work day in the format hh:mm.
  --home-office or -h: Whether the day is a home office day.
  --help: Show all avalible options
```

You can pass in either the `--day` option with a specific date or the `--day-str` option with a natural language date. If neither is specified, the script will prompt you for a default value of "yesterday".

You can provide the start and finish times of the work day using the `--start` and `--finish` options respectively. If you need to take a break during the work day, you can specify the start and finish times of the break using the --break-start and --break-finish options.

The `--home-office` option can be used to indicate whether the work day was done from home.

Once you have specified the required options, the script will use your Timetac access token to add the work day to your time tracking system. If your access token is not found, the script will prompt you to login and retrieve a new access token.

# Examples

To simply run interactively just invoke timetoc

```console
$ timetoc  
```

To add a work day for yesterday with a start time of 9:00 and a finish time of 17:00, you can run:

```console
$ timetoc  --day-str yesterday --start 9:00 --finish 17:00
```

To add a work day for a specific date, for example, March 15th, 2023, with a start time of 8:30, a finish time of 16:30, and a break from 12:00 to 12:30, you can run:

```console
$ timetoc  --day 2023-03-15 --start 8:30 --finish 16:30 --break-start 12:00 --break-finish 12:30
```

To add a work day for today that was done from home, you can run:

```console
$ timetoc  --day-str today --home-office
```


## Contributing

If you would like to contribute to the Timetoc, please fork the repository and submit a pull request. Contributions are always welcome!

## License

The Timetoc Work Tracker tool is licensed under the MIT License. See [LICENSE](https://github.com/guilhermeprokisch/timetoc/blob/main/LICENCE) for more information.


CHANGELOG.md
# Changelog

## [0.4.0](https://github.com/guilhermeprokisch/timetoc/compare/v0.3.2...v0.4.0) (2023-04-17)


### Features

* v0.4.0-beta ([ddc6f79](https://github.com/guilhermeprokisch/timetoc/commit/ddc6f794d458f2dcd69073c95b00ee87ff60fa05))


### Documentation

* readme.md ([d262650](https://github.com/guilhermeprokisch/timetoc/commit/d262650e36e3986754a20f640e4a79e758bb6975))
* updated readme ([6b89105](https://github.com/guilhermeprokisch/timetoc/commit/6b89105aa6c3e6acc4a2d937f8cfa37cea77c568))
* updated readme ([b861bf5](https://github.com/guilhermeprokisch/timetoc/commit/b861bf5459e269481f1e4f21f333532182d67744))
* updated readme ([616d9f1](https://github.com/guilhermeprokisch/timetoc/commit/616d9f1ebe035dc7a9ee5cbbc4ca3c3279ca2865))
* updated Readme.md ([83a1b87](https://github.com/guilhermeprokisch/timetoc/commit/83a1b876422ddf54c3ed11e8e26fb8a3461842d5))
* updated Readme.md ([02e9c1c](https://github.com/guilhermeprokisch/timetoc/commit/02e9c1cd1bc9933b1013b9743e68858775f33aef))

## [0.4.2](https://github.com/guilhermeprokisch/timetoc/compare/v0.4.1...v0.4.2) (2023-04-12)


### Documentation

* updated readme ([cb403d4](https://github.com/guilhermeprokisch/timetoc/commit/cb403d4ea2f3bd2e5fbdca244b36c29a75a471cb))
* updated Readme.md ([55561dd](https://github.com/guilhermeprokisch/timetoc/commit/55561dd873df816abbd6da83ae053fa567a9bdf5))

## [0.4.1](https://github.com/guilhermeprokisch/timetoc/compare/v0.4.0...v0.4.1) (2023-04-12)


### Documentation

* updated readme ([397ea9a](https://github.com/guilhermeprokisch/timetoc/commit/397ea9a1a62279ce8462bd7238b8ed8368d9a988))
* updated readme ([f78c0ef](https://github.com/guilhermeprokisch/timetoc/commit/f78c0ef9f077b616dbf68f0921317b41abeacaac))
* updated Readme.md ([83b4596](https://github.com/guilhermeprokisch/timetoc/commit/83b45967a75cfda9564dcf16fa18b4f1e0f97be1))

## [0.4.0](https://github.com/guilhermeprokisch/timetoc/compare/v0.3.2...v0.4.0) (2023-04-12)


### Features

* v1.0.0-beta ([0413ebe](https://github.com/guilhermeprokisch/timetoc/commit/0413ebe4d9fe818a6b79ff5edf349359428528a7))


### Documentation

* readme.md ([197f196](https://github.com/guilhermeprokisch/timetoc/commit/197f1963674fc28477d286731f2690316724ea6b))

## [0.3.2](https://github.com/guilhermeprokisch/timetoc/compare/v0.3.1...v0.3.2) (2023-04-02)


### Bug Fixes

* force bump version ([c99e40b](https://github.com/guilhermeprokisch/timetoc/commit/c99e40b4e532056d54544d5f42b3503c09d7b85e))
* release.yml syntax ([d396d98](https://github.com/guilhermeprokisch/timetoc/commit/d396d98412c0d2cd2d36b9f7bc91c3a66fb47d65))

## [0.3.1](https://github.com/guilhermeprokisch/timetoc/compare/v0.3.0...v0.3.1) (2023-04-02)


### Bug Fixes

* bump version ([a52021d](https://github.com/guilhermeprokisch/timetoc/commit/a52021d37cb0de1a897f2c52ca58bf8021742ad1))

## [0.3.0](https://github.com/guilhermeprokisch/timetoc/compare/v0.2.0...v0.3.0) (2023-04-02)


### Features

* run version ([a3f53ff](https://github.com/guilhermeprokisch/timetoc/commit/a3f53ff36f3b26e341af1ee1560127eb3683b41e))


### Bug Fixes

* **ci:** pr_lint format ([0836a5c](https://github.com/guilhermeprokisch/timetoc/commit/0836a5c9597dad177616760965cf230a4841f707))


### Documentation

* setup mkdocs tasks and config ([8f46e92](https://github.com/guilhermeprokisch/timetoc/commit/8f46e92d283093531427702c765d08264b866492))

## [0.2.0](https://github.com/guilhermeprokisch/timetoc/compare/v0.1.0...v0.2.0) (2023-04-02)


### Features

* add message as parameter to main ([4977545](https://github.com/guilhermeprokisch/timetoc/commit/4977545521742c035e4a68a6a6b78ffc1fbbcedf))

## 0.1.0 (2023-04-02)


### Features

* add main ([8aa857a](https://github.com/guilhermeprokisch/timetoc/commit/8aa857aa32bd0b2f29da6e03acd30e57b77b3eac))


src/timetoc/login.py
import contextlib
import json
import os
import time

import keyring
import selenium.webdriver as webdriver
import selenium.webdriver.support.ui as ui
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
from seleniumwire import webdriver

TIME_TRACK_BASE_URL = os.environ["TIME_TRACK_BASE_URL"]


def get_access_token(email=None, password=None, headless=False):
    if email and password:
        chrome_options = Options()
        chrome_options.add_argument("--headless")
        drive = webdriver.Chrome(
            options=chrome_options, seleniumwire_options={"disable_encoding": True}
        )
    else:
        chrome_options = Options()
        if headless:
            chrome_options.add_argument("--headless")
        drive = webdriver.Chrome(
            seleniumwire_options={"disable_encoding": True}, options=chrome_options
        )

    with contextlib.closing(drive) as driver:
        driver.get(TIME_TRACK_BASE_URL)
        if email and password:
            wait_email = ui.WebDriverWait(driver, 60)  # timeout after 120 seconds
            wait_email.until(
                lambda driver: driver.find_element(By.CSS_SELECTOR, "#i0116")
            )
            email_input = driver.find_element(By.CSS_SELECTOR, "#i0116")
            email_input.send_keys(email)
            email_input.send_keys(Keys.ENTER)

            wait_password = ui.WebDriverWait(driver, 60)  # timeout after 120 seconds
            wait_password.until(
                lambda driver: driver.find_element(By.CSS_SELECTOR, "#i0118")
            )
            password_input = driver.find_element(By.CSS_SELECTOR, "#i0118")
            password_input.send_keys(password)
            password_input.send_keys(Keys.ENTER)
            password_input.send_keys(Keys.ENTER)

            time.sleep(2)
            sign_in = driver.find_element(By.CSS_SELECTOR, "#idSIButton9")
            sign_in.click()

            time.sleep(2)
            code = driver.find_element(By.CSS_SELECTOR, "#idRichContext_DisplaySign")
            print("Insert this code in the Authenticator App 👉", code.text)

        wait_login = ui.WebDriverWait(driver, 1000)  # timeout after 120 seconds
        wait_login.until(
            lambda driver: driver.find_element(By.ID, "panel-1010-innerCt")
        )
        for request in driver.requests:
            if request.response:
                if "load_all_ext" in request.url:
                    body = request.response.body
                    data = json.loads(str(body, "utf-8"))
                    access_token = data["userData"]["accessToken"]
                    keyring.set_password("system", "timetoc_access_token", access_token)
                    print("🔑 Access token obtained and saved!")
                    return access_token


src/timetoc/timeparser.py
import datetime

import holidays
from dateutil.relativedelta import relativedelta


def parse_date_range(phrase, exclude_holidays=True, exclude_weekends=True):
    """
    Returns a list of datetime objects representing a date range based on a phrase describing the range.

    Parameters:
        phrase (str): A string describing the date range.
        exclude_holidays (bool): Whether to exclude Austrian public holidays from the resulting list of dates. Default is False.
        exclude_weekends (bool): Whether to exclude weekends (Saturdays and Sundays) from the resulting list of dates. Default is False.

    Returns:
        list: A list of datetime objects representing the dates in the range.
    """
    now = datetime.datetime.now().replace(hour=0, minute=0, second=0, microsecond=0)

    if phrase == "today":
        start_date = now
        end_date = now + datetime.timedelta(days=1)
    elif phrase == "yesterday":
        start_date = now - datetime.timedelta(days=1)
        end_date = now
    elif phrase == "tomorrow":
        start_date = now + datetime.timedelta(days=1)
        end_date = start_date + datetime.timedelta(days=1)
    elif phrase == "this week":
        start_date = now - datetime.timedelta(days=now.weekday())
        end_date = start_date + datetime.timedelta(weeks=1)
    elif phrase == "last week":
        end_date = now - datetime.timedelta(days=now.weekday())
        start_date = end_date - datetime.timedelta(weeks=1)
    elif phrase == "next week":
        start_date = now + datetime.timedelta(days=(7 - now.weekday()))
        end_date = start_date + datetime.timedelta(weeks=1)
    elif phrase == "this month":
        start_date = now.replace(day=1)
        end_date = (start_date + relativedelta(months=1)).replace(day=1)
    elif phrase == "last month":
        end_date = now.replace(day=1)
        start_date = (end_date - relativedelta(months=1)).replace(day=1)

    date_range = []
    delta = datetime.timedelta(days=1)
    while start_date < end_date:
        if not exclude_weekends or start_date.weekday() not in [5, 6]:
            date_range.append(start_date)
        start_date += delta

    if exclude_holidays:
        at_holidays = holidays.AT(years=start_date.year)
        date_range = [
            d.strftime("%Y-%m-%d") for d in date_range if d not in at_holidays
        ]

    return date_range


src/timetoc/timetracking.py
import os
from datetime import datetime

import requests

TIME_TRACK_BASE_URL = os.environ["TIME_TRACK_BASE_URL"]


def make_headers(token=None):
    return {
        "Authorization": f"Bearer {token}",
    }


def add_time_entry(
    day="2023-03-14",
    start_time="09:10",
    end_time="12:00",
    break_time=False,
    token=None,
):
    headers = make_headers(token)

    task_id = 4
    if break_time:
        task_id = 9

    data = {
        "0": {
            "user_id": 556,
            "task_id": task_id,
            "start_time_timezone": "Europe/Vienna",
            "end_time_timezone": "Europe/Vienna",
            "start_time": f"{day} {start_time}:00",
            "end_time": f"{day} {end_time}:00",
            "notes": "",
            "start_type_id": "0",
            "end_type_id": "0",
        },
    }

    response = requests.post(
        f"{TIME_TRACK_BASE_URL}/userapi/v3/timeTrackings/create/",
        headers=headers,
        json=data,
    )

    response.raise_for_status()

    if response.status_code == 200:
        if break_time:
            print(
                f"✅ Break time entry starting at {start_time} ending at {end_time}  add successfully ☕ "
            )
            return

        print(
            f"✅ Time entry starting at {start_time} ending at {end_time}  add successfully 🕒"
        )


def add_home_office(day, token):
    headers = make_headers(token)
    data = {
        "user_id": "556",
        "subtype_id": "12",
        "type_id": "3",
        "from_date": f"{day}",
        "to_date": f"{day}",
        "request_duration_entity": "d",
        "duration": "1",
        "request_comment": "",
    }

    response_home_office = requests.post(
        f"{TIME_TRACK_BASE_URL}/userapi/v3/absences/create/",
        headers=headers,
        data=data,
    )
    if response_home_office.status_code == 200:
        print("✅ Home office entry add successfully 🏠")


def add_work_day(day, start, finish, break_start, break_finish, is_home_office, token):
    add_time_entry(day, start_time=start, end_time=break_start, token=token)
    add_time_entry(
        day, start_time=break_start, end_time=break_finish, break_time=True, token=token
    )
    add_time_entry(day, start_time=break_finish, end_time=finish, token=token)
    if is_home_office:
        add_home_office(day, token)

    work_time = total_work_time(start, finish, break_start, break_finish)
    print(f"Total work time on {day} -> {work_time} hours ")


def hours_difference(start, end):
    t1 = datetime.strptime(start, "%H:%M")
    t2 = datetime.strptime(end, "%H:%M")
    return t2 - t1


def total_work_time(start, finish, break_start, break_finish):
    working_time_1 = hours_difference(start, break_start)
    working_time_2 = hours_difference(break_finish, finish)
    total_work_time = working_time_1 + working_time_2
    hours_diff = "{:0>8}".format(str(total_work_time))
    return hours_diff


src/timetoc/cli.py
"""A command-line tool for tracking work time in timetoc.

This module provides functions to add work days and get the corresponding time tracking 
information using the timetoc API. It uses the typer module for the command-line interface, 
keyring for storing login information securely, and rich for pretty-printing information.

Functions:
    access_token_by_login: Retrieves access token from timetoc by logging in and 
        authenticating the user.
    get_day_info: Retrieves user input for work day information such as start and finish time,
        break start and finish time, whether the day is a home office day, and the day's date.
    add_work_day_with_retry: Adds a work day to the timetoc server with a retry mechanism in 
        case the authentication token has expired.
    main: The main function that runs the command-line interface.
"""

from datetime import datetime

import keyring
import typer
from rich.console import Console
from rich.prompt import Prompt

from timetoc.login import get_access_token
from timetoc.timeparser import parse_date_range
from timetoc.timetracking import add_work_day

app = typer.Typer()
console = Console()

# Constants
HEADLESS_LOGIN_DEFAULT = False
START_TIME_DEFAULT = "09:00"
FINISH_TIME_DEFAULT = "17:30"
BREAK_START_DEFAULT = "12:30"
BREAK_FINISH_DEFAULT = "13:00"
HOMEOFFICE_DAYS_DEFAULT = ["Tuesday", "Thursday"]


def access_token_by_login():
    """Retrieve access token from timetoc by logging in and authenticating the user.

    The function first prompts the user to select whether to use a headless login or not. If
    headless login is selected, the function retrieves the user's email and password from the
    system's keyring. If the email and password are not stored, the function prompts the user to
    enter them and saves them to the keyring if the user chooses to do so. The function then calls
    the get_access_token function to retrieve the access token using the provided email, password,
    and headless parameters.

    Returns:
        str: The retrieved access token.
    """

    headless_login = typer.confirm("Headless login?", default=HEADLESS_LOGIN_DEFAULT)

    if headless_login:
        email = keyring.get_password("system", "timetoc_email")
        password = keyring.get_password("system", "timetoc_password")

        if not email or not password:
            email = Prompt.ask("Email")
            password = Prompt.ask("Password", password=True)
            save_pass = typer.confirm("Save Login and Password?", default=True)

            if save_pass:
                keyring.set_password("system", "timetoc_email", email)
                keyring.set_password("system", "timetoc_password", password)

        access_token = get_access_token(email=email, password=password, headless=True)
    else:
        access_token = get_access_token(headless=False)

    return access_token


def get_additional_info(
    day,
    start,
    finish,
    break_start,
    break_finish,
    is_home_office,
):
    """Retrieve user input for work day information.

    The function prompts the user to enter the start and finish time, break start and finish time,
    and whether the day is a home office day. The function then returns a tuple containing the
    start and finish time, break start and finish time, the week day name of the day (e.g.,
    Monday, Tuesday, etc.), and whether the day is a home office day.

    Args:
        day (str): The date of the work day in the format yyyy-mm-dd.

    Returns:
        Tuple[str, str, str, str, str, bool]: A tuple containing the start and finish time, break
        start and finish time, the week day name of the day, and whether the day is a home office
        day.
    """

    day_datetime = datetime.strptime(day, "%Y-%m-%d")
    week_day_name = day_datetime.strftime("%A")
    if not start:
        start = Prompt.ask("Start Time", default=START_TIME_DEFAULT)
    if not finish:
        finish = Prompt.ask("Finish Time", default=FINISH_TIME_DEFAULT)
    if not break_start:
        break_start = Prompt.ask("Break Start", default=BREAK_START_DEFAULT)
    if not break_finish:
        break_finish = Prompt.ask("Break Finish", default=BREAK_FINISH_DEFAULT)
    if not is_home_office:
        is_home_office = False if week_day_name in HOMEOFFICE_DAYS_DEFAULT else True
        is_home_office = typer.confirm("Is Home Office?", default=is_home_office)

    return start, finish, break_start, break_finish, week_day_name, is_home_office


def add_work_day_with_retry(
    day, start, finish, break_start, break_finish, is_home_office, access_token
):
    """Add a work day to the timetoc server with a retry mechanism in case the authentication
    token has expired.

    The function first calls the add_work_day function to add the work day with the provided
    parameters and access token. If an exception is raised due to an expired authentication token,
    the function retrieves a new access token by calling the access_token_by_login function and
    then calls add_work_day again with the new access token.

    Args:
        day (str): The date of the work day in the format yyyy-mm-dd.
        start (str): The start time of the work day in the format hh:mm.
        finish (str): The finish time of the work day in the format hh:mm.
        break_start (str): The start time of the break in the work day in the format hh:mm.
        break_finish (str): The finish time of the break in the work day in the format hh:mm.
        is_home_office (bool): Whether the day is a home office day or not.
        access_token (str): The authentication token to access the timetoc API.
    """

    try:
        add_work_day(
            day=day,
            start=start,
            finish=finish,
            break_start=break_start,
            break_finish=break_finish,
            is_home_office=is_home_office,
            token=access_token,
        )
    except:
        console.print("❌ Token expired please login again")
        access_token = access_token_by_login()
        add_work_day(
            day=day,
            start=start,
            finish=finish,
            break_start=break_start,
            break_finish=break_finish,
            is_home_office=is_home_office,
            token=access_token,
        )


@app.command()
def main(
    day: str = typer.Option(
        None, "--day", "-d", help="The date of the work day in the format yyyy-mm-dd."
    ),
    day_str: str = typer.Option(
        None,
        "--day-str",
        "-ds",
        help="The date of the work day in natural languge. E.g 'today, yesterday'",
    ),
    start: str = typer.Option(
        None,
        "--start",
        "-s",
        help="The start time of the work day in the format hh:mm.",
    ),
    finish: str = typer.Option(
        None,
        "--finish",
        "-f",
        help="The finish time of the work day in the format hh:mm.",
    ),
    break_start: str = typer.Option(
        None,
        "--break-start",
        "-bs",
        help="The start time of the break in the work day in the format hh:mm.",
    ),
    break_finish: str = typer.Option(
        None,
        "--break-finish",
        "-bf",
        help="The finish time of the break in the work day in the format hh:mm.",
    ),
    is_home_office: bool = typer.Option(
        None,
        "--home-office",
        "-h",
        help="Whether the day is a home office day.",
    ),
):
    access_token = keyring.get_password("system", "timetoc_access_token")

    if not access_token:
        console.print("Access token not found!", style="red")
        access_token = access_token_by_login()

    days = []

    if not day_str and not day:
        day_str = Prompt.ask("Day String", default="yesterday")

    if not day and day_str:
        days = parse_date_range(day_str)

    if day and not day_str:
        days = [day]

    for day in days:
        (
            start,
            finish,
            break_start,
            break_finish,
            week_day_name,
            is_home_office,
        ) = get_additional_info(
            day, start, finish, break_start, break_finish, is_home_office
        )

        console.print("Day: ", f"{day} {week_day_name}", style="bold red")

        add_work_day_with_retry(
            day=day,
            start=start,
            finish=finish,
            break_start=break_start,
            break_finish=break_finish,
            is_home_office=is_home_office,
            access_token=access_token,
        )


if __name__ == "__main__":
    app()
